.github/workflows/build-and-test.yml

```diff
@@ -1,415 +1,540 @@
-
-            - name: Install Node.js
-              uses: actions/setup-node@v2
-              with:
-                  node-version: 14.x
-
-            - name: Install Workbench Dependencies
-              working-directory: ./workbench
-              run: |
-                  yarn config set network-timeout 600000 -g
-                  yarn install
-
-            - name: run workbench tests
-              shell: bash -l {0}
-              working-directory: ./workbench
-              env:
-                  CI: true
-              run: yarn test
+name: Run Tests and Build
+
+on: [push, pull_request]
+
+defaults:
+  run:
+    shell: bash -l {0}
+
+concurrency:
+  # make sure only one run of this workflow for a given PR or a given branch
+  # can happen at one time. previous queued or started runs will be cancelled.
+  # github.workflow is the workflow name
+  # github.ref is the ref that triggered the workflow run
+  # on push, this is refs/heads/<branch name>
+  # on pull request, this is refs/pull/<pull request number>/merge
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+env:
+  # build: dependency of make install
+  # nomkl: make sure numpy w/out mkl
+  # setuptools_scm: needed for versioning to work
+  DEFAULT_DEPENDENCIES: build nomkl setuptools_scm
+  PYTHON_VERSION: 3.9
+
+jobs:
+  check-syntax-errors:
+    name: Check for syntax errors
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up python
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      # this is fast enough that it's not worth caching
+      - name: Set up environment
+        run: |
+          echo ${{ github.workflow }}
+          echo ${{ github.ref }}
+          echo ${{ github.head_ref }}
+          pip install flake8
+
+      - name: Lint with flake8
+        run: |
+          # stop the build if there are Python syntax errors or undefined names
+          python -m flake8 src --count --select=E9,F63,F7,F82 --show-source --statistics
+          # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+          python -m flake8 src --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+
+  run-model-tests:
+    name: Run model tests
+    runs-on: ${{ matrix.os }}
+    needs: check-syntax-errors
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: [3.7, 3.8, 3.9]
+        os: [windows-latest, macos-latest]
+        include:  # test with a range of numpy and gdal versions
+          - python-version: 3.7
+            numpy: "numpy==1.16.6"  # oldest version compatible with other requirements
+            gdal: "gdal==3.1.2"
+          - python-version: 3.8
+            numpy: "numpy==1.20.0"  # in between version
+            gdal: "gdal==3.3.1"
+          - python-version: 3.9
+            numpy: numpy           # latest version
+            gdal: gdal

```

Comment:
I sort of remember pinning them partly to test with a wider range of versions - to make sure the full range of versions we claim to support actually works. Maybe there was some issue that only happened with a particular GDAL and numpy version together? But we can't actually test every combination of dependency versions, and I'm not sure it's our responsibility to do so. So I'd be happy to drop the numpy and gdal version pins.

> GDAL makes sense to me though to keep in the matrix ... that's a whole lot easier to install from conda-forge than to rely on pip grabbing the latest version.

I think it will behave the same without specifying it in the matrix?

Reply:
> I think it will behave the same without specifying it in the matrix?

It looks like maybe the conda-forge GDAL builds have caught up, so on python3.7, I'm getting a conda-forge build of GDAL 3.4.1 instead of the matrix's GDAL 3.1.2.

Yeah, we can't actually test every combination of dependency versions and I fully agree it shouldn't be our responsibility to do so.  I'm OK with keeping the GDAL pins for the sake of testing a few different GDAL versions, but also OK with removing them.